@helia/verified-fetch announcement (#683)
ipfs/ipfs-blog: src/_blog/verified-fetch.md

---
date: 2024-04-11
permalink: /verified-fetch/
title: 'Verified IPFS Retrieval in Browsers with @helia/verified-fetch'

> description: 'Verified Fetch is library streamlining verified retrieval of IPFS content in browsers and JS runtimes, with native support for IPNS, and DNSLink resolution'

Reviewer: ```suggestion
description: 'Verified Fetch is a library for streamlining verified retrieval of IPFS content in browsers and JS runtimes, with native support for IPNS, and DNSLink resolution'
```

Trustless IPFS Gateways are useful for browsers because they can be composed in a way that unleashes many of the aforementioned benefits of IPFS and content addressing.

> **Note:** Browser constraints prevent you from opening connections to "random" addresses that don't have a CA signed certificate, making it hard to build IPFS clients for browsers that go straight to providers. Newer transport such as [WebTransport](https://docs.libp2p.io/concepts/transports/webtransport/) and [WebRTC-direct](https://docs.libp2p.io/concepts/transports/webrtc/) address this challenge in a way that may be able to reduce dependency on IPFS Gateways in the future.

## Solution: Verified Fetch

> [`@helia/verified-fetch`](https://www.npmjs.com/package/@helia/verified-fetch) or simply **Verified Fetch** is a new JavaScript library by [The Shipyard team](https://blog.ipfs.tech/shipyard-hello-world/) that makes verified IPFS retrieval from trustless gateways easy. It's written in TypeScript with Web APIs so you can run it on any modern JS runtime.

Reviewer: > can run it on any modern JS runtime

I think there are still issues with Deno running libp2p under the hood, but verified-fetch with helia/http should work in deno. idk if we have tests for this...

Author: I'll give Deno a try to make sure. I'll change the phrasing a bit to. Either way this remains a goal for us as we even advertise in the main Helia README https://github.com/ipfs/helia/blob/main/README.md?plain=1#L12

  See the Pen <a href="https://codepen.io/2color/pen/ExJdayy">
  Fetching a CID with @helia/verified-fetch example</a> by Daniel Norman (<a href="https://codepen.io/2color">@2color</a>)
</iframe>

Finally, since you can store **any** kind of file with UnixFS, if you want the `Content-Type` header of the `Response` object to be sniffed on the [Magic Bytes of the retrieved binary data](https://en.wikipedia.org/wiki/Content_sniffing), you can pass the `contentTypeParser` option as follows:

<iframe height="300" style="width: 100%;" scrolling="no" title="Content-Type sniffing with contentTypeParser @helia/verified-fetch example" src="https://codepen.io/2color/embed/JjVmoLy?default-tab=js%2Cresult&editable=true" frameborder="no" loading="lazy" allowtransparency="true" allowfullscreen="true">
  See the Pen <a href="https://codepen.io/2color/pen/JjVmoLy">
  Content-Type sniffing with contentTypeParser @helia/verified-fetch example</a> by Daniel Norman (<a href="https://codepen.io/2color">@2color</a>)
</iframe>

### Customizable

> By default, Verified Fetch uses [`@helia/http`](https://github.com/ipfs/helia/tree/main/packages/http#heliahttp): a lightweight version of Helia on IPFS over HTTP with Trustless Gateways. However, you can [pass an instance of Helia that is customized to your needs](https://github.com/ipfs/helia-verified-fetch/tree/main/packages/verified-fetch#usage-with-customized-helia). A common use-case might be when running on Node.js where you might want to lean more heavily on direct retrieval using Bitswap over TCP.

Reviewer: one more comment @2color.. doesn't seem clear to chatGpt summarizer that p2p functionality is also supported. I think we should make it more clear that direct p2p retrieval is supported with custom helia nodes (and helia/http's use of delegated routing)

<img width="732" alt="image" src="https://github.com/ipfs/ipfs-blog/assets/1173416/15874ed7-08b5-4093-bd10-4cbe715f7873">

Author: Good point. Added it explicitly twice in the blog post including an example with the normal helia that comes with libp2p.

---
date: 2024-04-11
permalink: /verified-fetch/
title: 'Verified IPFS Retrieval in Browsers with @helia/verified-fetch'
description: ''
author: Daniel Norman
header_image: /dapps-ipfs/header.png
tags:
  - ipfs
  - dapps
  - Helia
  - ipns
  - ens
---

## Announcing @helia/verified-fetch

> The Shipyard team is thrilled to announce **`@helia/verified-fetch`** is now ready for broader adoption. Verified Fetch is a [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)-like library streamlining verified retrieval of IPFS content in browsers and JS runtimes, with native support IPNS, and DNSLink resolution. [Try it out](https://www.npmjs.com/package/@helia/verified-fetch) and let us know what you think.

Reviewer 2: ```suggestion
The Shipyard team is thrilled to announce **`@helia/verified-fetch`** is now ready for broader adoption. Verified Fetch is a [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)-like library streamlining verified retrieval of IPFS content in browsers and JS runtimes, with native support for IPNS, and DNSLink resolution. [Try it out](https://www.npmjs.com/package/@helia/verified-fetch) and let us know what you think.
```

> This blog post covers the challenges of IPFS retrieval in browsers and how @helia/verified-fetch addresses them with runable examples. Feel free to jump ahead to [Solution: Verified Fetch](#solution-verified-fetch)

Reviewer 2: ```suggestion
This blog post covers the challenges of IPFS retrieval in browsers and how `@helia/verified-fetch` addresses them with runable examples. Feel free to jump ahead to [Solution: Verified Fetch](#solution-verified-fetch)
```

### IPFS Gateways are a useful abstraction for IPFS retrieval

Trusted IPFS Gateways abstract much of the complexity (peer-to-peer connectivity, content routing, content retrieval, verification) of IPFS with a straightforward HTTP API to the IPFS network:

<img alt="gateway architecture diagram" src="../assets/verified-fetch/gateways.png" width="500">

The beauty of IPFS Gateways is in how simple they are to use: you append the CID to the URL of the Gateway and all IPFS magic is handled by the gateway for you.

For example, fetching an image with the CID: [`bafk...beom`](https://cid.ipfs.tech/#bafkreie7ohywtosou76tasm7j63yigtzxe7d5zqus4zu3j6oltvgtibeom) is as simple as constructing the URL: `https://ipfs.io/ipfs/bafkreie7ohywtosou76tasm7j63yigtzxe7d5zqus4zu3j6oltvgtibeom` which can be passed to the `src` attribute of an `<img>`, as follows:

<img class="py-4" alt="image loaded from an IPFS gateway" src="https://ipfs.io/ipfs/bafkreie7ohywtosou76tasm7j63yigtzxe7d5zqus4zu3j6oltvgtibeom" width="350">

### Trusting an IPFS Gateway without verifying is an anti-pattern

> **Nonetheless, fetching from a trusted IPFS gateway without verifying is an anti-pattern** and goes against [the principles of IPFS](https://specs.ipfs.tech/architecture/principles/#verification-matters).

Reviewer 2: Since we have "Implementations MAY relax this requirement in controlled environments" in https://specs.ipfs.tech/architecture/principles/#ipfs-implementation-requirements perhaps a stronger point here is to use "third-party"?
```suggestion
**Nonetheless, fetching from a third-party IPFS gateway without verifying is an anti-pattern** and goes against [the principles of IPFS](https://specs.ipfs.tech/architecture/principles/#verification-matters).
```

Content addressing in IPFS frees you from the model of a single canonical source for data. This is a powerful concept and the root of IPFS' benefits: resilience, censorship resistance, and trustlessness. But, **fully reaping the benefits of IPFS requires verification**.

### Verification facilitates resilience multi-source retrieval

Verifying IPFS content as part of the retrieval process allows you to fetch it from multiple sources –either providers or gateways– without trusting them, because verification ensures the integrity of the data.

This comes with the downstream benefit of resilience: if one provider or gateway is unavailable, unreachable, or censored, you can still retrieve the CID from another (as long as other providers are available).

### Trustless IPFS Gateways enable verification in browsers

[Trustless IPFS Gateways](https://specs.ipfs.tech/http-gateways/trustless-gateway/) have been gaining steam as a means of enabling verification and its downstream benefits with the simplicity of IPFS Gateways over HTTP.

> Trustless IPFS Gateways' response types are fully and incrementally verifiable: clients can decide between a [raw block](https://docs.ipfs.tech/concepts/glossary/#block) or a [CAR stream](https://docs.ipfs.tech/concepts/glossary/#car).

Reviewer 2: :thought_balloon: good opportunity to put IANA content types in front of the reader (or at least for SEO)

```suggestion
Trustless IPFS Gateways' response types are [fully and incrementally verifiable](https://docs.ipfs.tech/reference/http/gateway/#trustless-verifiable-retrieval): clients can decide between a [raw block](https://docs.ipfs.tech/concepts/glossary/#block) ([`application/vnd.ipld.raw`](https://www.iana.org/assignments/media-types/application/vnd.ipld.raw)) or a [CAR stream](https://docs.ipfs.tech/concepts/glossary/#car) ([`application/vnd.ipld.car`](https://www.iana.org/assignments/media-types/application/vnd.ipld.car)).
```

## Solution: Verified Fetch

[`@helia/verified-fetch`](https://www.npmjs.com/package/@helia/verified-fetch) or simply **Verified Fetch** is a new library by [The Shipyard team](https://blog.ipfs.tech/shipyard-hello-world/) that makes verified retrieval from trustless gateways easy.

It's the culmination of multiple streams of work we undertook to bring seamless and deeper IPFS integrations to browsers and improve the development experience with IPFS.

<!-- These efforts include:
- [Delegated routing over HTTP](https://specs.ipfs.tech/routing/http-routing-v1/) and [hosted someguy](https://docs.ipfs.tech/concepts/public-utilities/#delegated-routing)
- [Trustless Gateway](https://specs.ipfs.tech/http-gateways/trustless-gateway/)
- [Helia](https://github.com/ipfs/helia/): lean, modular, and modern TypeScript implementation of IPFS for the prolific JS and browser environments.
- [Helia/http](https://github.com/ipfs/helia/tree/main/packages/http): A lightweight version of Helia on IPFS over HTTP with Trustless Gateways -->

### Familiar, like the Fetch API

> Verified Fetch is modeled after the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) and returns [Response object](https://developer.mozilla.org/en-US/docs/Web/API/Response) making it easy to adopt and reason about.

Reviewer 2: ```suggestion
Verified Fetch is modeled after the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) and returns [`Response` object](https://developer.mozilla.org/en-US/docs/Web/API/Response) making it easy to adopt and reason about.
```

```ts
import { createVerifiedFetch } from '@helia/verified-fetch'
import { dnsJsonOverHttps } from '@multiformats/dns/resolvers'

const verifiedFetch = await createVerifiedFetch({
  dnsResolvers: {
    'eth.': dnsJsonOverHttps('https://dns.eth.limo/dns-query'),
    '.': dnsJsonOverHttps('https://cloudflare-dns.com/dns-query')
  }
})
const resp = await verifiedFetch('ipns://vitalik.eth/images/scaling-files/cryptokitties.png')
```

The following example uses Verified Fetch to fetch and verify an image from Vitalik's website:

<iframe height="500" style="width: 100%;" scrolling="no" title="Resolving and Fetching a DNSLink with @helia/verified-fetch and custom DoH endpoint example" src="https://codepen.io/2color/embed/wvZXRPa?default-tab=js%2Cresult" frameborder="no" loading="lazy" allowtransparency="true" allowfullscreen="true">
  See the Pen <a href="https://codepen.io/2color/pen/wvZXRPa">
  Resolving and Fetching a DNSLink with @helia/verified-fetch and custom DoH endpoint example</a> by Daniel Norman (<a href="https://codepen.io/2color">@2color</a>)
  on <a href="https://codepen.io">CodePen</a>.
</iframe>

### Supports a wide range of data types

> As you may have noticed, you can use Verified Fetch to fetch a wide range of data types. Verified Fetch abstracts much of the complexity of IPLD codecs, supporting [UnixFS](https://docs.ipfs.tech/concepts/file-systems/#unix-file-system-unixfs), dag-cbor, and dag-json out of the box. This frees you to focus on your application. The `text()`, `.blob()`, and .`arrayBuffer()` methods will work as expected without a detailed content type.

Reviewer 2: ```suggestion
As you may have noticed, you can use Verified Fetch to fetch a wide range of data types. Verified Fetch abstracts much of the complexity of IPLD codecs, supporting [UnixFS](https://docs.ipfs.tech/concepts/file-systems/#unix-file-system-unixfs), [dag-cbor](https://ipld.io/specs/codecs/dag-cbor/), and [dag-json](https://ipld.io/specs/codecs/dag-json/) out of the box. This frees you to focus on your application. The `text()`, `.blob()`, and .`arrayBuffer()` methods will work as expected without a detailed content type.
```

> By default, if the response can be parsed as JSON, Verified Fetch sets the `Content-Type` header of the Response object to as `application/json`, otherwise it sets it as application/octet-stream.

Reviewer 2: ```suggestion
By default, if the response can be parsed as JSON, Verified Fetch sets the `Content-Type` header of the Response object to as `application/json`, otherwise it sets it as `application/octet-stream`.
```

[You can also pass the `Accept` header](https://github.com/ipfs/helia-verified-fetch/tree/main/packages/verified-fetch#the-accept-header) to override certain response processing, or to ensure that the final Content-Type of the response is the one that is expected.

### Customizable

> By default, Verified Fetch uses [Helia/http](https://github.com/ipfs/helia/tree/main/packages/http): a lightweight version of Helia on IPFS over HTTP with Trustless Gateways. However, you can pass an instance of Helia that is customized to your needs. A common use-case might be when running on Node.js where you might want to lean more heavily on direct retrieval using Bitswap over TCP.

Reviewer 2: ```suggestion
By default, Verified Fetch uses [`@helia/http`](https://github.com/ipfs/helia/tree/main/packages/http#heliahttp): a lightweight version of Helia on IPFS over HTTP with Trustless Gateways. However, you can pass an instance of Helia that is customized to your needs. A common use-case might be when running on Node.js where you might want to lean more heavily on direct retrieval using Bitswap over TCP.
```
